### PR TITLE
Support direct export of block-wise FP8 weights and scaling factors

### PR DESCRIPTION
**Pull request by eternally-z**
# What does this PR do ?

This PR adds support for directly exporting block-wise **FP8 weights** and their corresponding inverse scales (`scale_inv`) when calling `export_hf_weights`.

### Motivation
In scenarios such as LLM Reinforcement Learning with FP8 training and rollout, the current workflow typically requires de-quantizing FP8 weights back to BF16 during export, only for them to be re-quantized during the rollout phase.

This "de-quantize -> re-quantize" process introduces:
1.  Redundant computational overhead.
2.  Potential precision loss due to repeated conversion.
3.  Increased transfer size: Exporting as BF16 consumes 2x the storage compared to the native FP8 format.

By allowing direct export of FP8 weights and scales, we enable a seamless and efficient pipeline from training to rollout.

### Implementation Details
1.  `export_hf_weights`:
    - Introduced a new function `build_export_fp8_tasks`.
    - this function allows the exporter to dump raw FP8 integer weights and `scale_inv` factors directly, bypassing the conversion to BF16.
    
2.  `load_hf_weights`:
    - Modified the loading process to retain the original `state_dict` when `fp8_param_gather` is enabled. 
    - This ensures that Megatron initializes the model using the **original BF16 weights**, rather than using the de-quantized results from FP8 parameters .

### Scope & Limitations
- Supported: Dense models & MoE models.
- Verification: Verified on Qwen3-8B (Dense) and Qwen3-30B-A3B (MoE).

@AniZpZ
# Changelog

- Add specific line by line info of high level changes in this PR.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci)in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

## Summary by CodeRabbit

* **New Features**
  * Introduced a new weight export format option providing enhanced control over model precision during conversion.
  * Added support for adapter weight conversion during model merging operations.

* **Improvements**
  * Enhanced weight conversion with improved handling of quantized parameters and scale data.
  * Strengthened QKV weight tensor handling to support diverse weight configurations during model transformation.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/megatron/bridge/models/conversion/param_mapping.py [hunk 1]
```diff
@@ -2292,8 +2292,37 @@ def split_qkv_weights(
         hidden_size = 1
         qkv_reshaped = qkv.view(qkv_total_dim, head_size)
     else:
-        hidden_size = qkv.shape[-1]
-        qkv_reshaped = qkv.view(qkv_total_dim, head_size, hidden_size)
+        # NOTE: For standard (BF16/FP16) weights, `head_size` is the usual kv_channels/head_dim.
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/param_mapping.py:2587]: will fp8 affect other module layers, other than qkv?

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 2]
```diff
@@ -1276,6 +1335,242 @@ def build_conversion_tasks(
...
+        for d in fp8_flags_list:
+            if not d:
+                continue
+            for k, v in d.items():
+                if v:
+                    global_fp8_flags[k] = True
+
+        return global_fp8_flags
+
+    def build_export_fp8_tasks(
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/model_bridge.py:1565]: is this only for export, import will not work

File: src/megatron/bridge/models/conversion/param_mapping.py (hunk 1 above)
Comment (eternally-z) [reply src/megatron/bridge/models/conversion/param_mapping.py:2587]: Yes, besides QKV, we found that gdn_linear_weights and kv_weights are also affected.
The root cause is similar: the original code reshapes tensors using standard model dimensions (e.g., `qkvz = qkvz.reshape(-1, hidden_size)`). This logic fails for FP8 scale parameters because their shapes are determined by the block size (scaled down), causing dimension mismatches.
We verified this on Qwen3-8B and Qwen3-30B-A3B. For now, we have temporarily only updated the split logic for the QKV module to accommodate the FP8 scale dimensions, and we can easily apply this to gdn_linear_weights and kv_weights.

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 2 above)
Comment (eternally-z) [reply src/megatron/bridge/models/conversion/model_bridge.py:1565]: Yes, this PR specifically targets the export of block-wise FP8 weights and scales. Importing FP8 weights/scales is not supported at this stage.

**Comment by yaoyu-33** [comment]
@eternally-z: The lint is failing, plz check Contribute.md, need to run pre-commit

**Comment by eternally-z** [comment]
> /ok to test [fd9ba1d](https://github.com/NVIDIA-NeMo/Megatron-Bridge/commit/fd9ba1dd54b443984ef017e1d899d58fa4dddfc8)

@yaoyu-33  hi, we updated the access method for unquantized_state_dict.
Previously, we added it to the return values of load_weights_hf_to_megatron. However, this caused tests to fail. This made us realize that modifying the function's return signature breaks the tests and risks breaking other downstream usages.
To fix this, it is now stored and accessed as an attribute of the model_bridge instance instead.

**Comment by eternally-z** [comment]
@yaoyu-33 hi, I noticed the CI passed but the codecov/patch failed. I'm not sure how to fix the problem. Should I add the tests myself?

**Comment by gautham-kollu** [comment]
@eternally-z There is a codecov failure and merge conflict. Can you please resolve that ?

**Comment by eternally-z** [comment]
> @eternally-z There is a codecov failure and merge conflict. Can you please resolve that ?

@gautham-kollu Yes, I noticed that. I have already resolved the merge conflict.
However, I'm not  sure how to handle the codecov/patch failure. It shows that the test coverage for my newly added code is insufficient. How should I go about fixing this? Should I add the tests myself?

**Comment by cuichenx** [comment]
@eternally-z can you please add a unit test for fp8 export?

**Comment by eternally-z** [comment]
> @eternally-z can you please add a unit test for fp8 export?

I've added a unit test for FP8 export. Please take a look.

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1994/commits/b32e0c0cc6a2f03d0c9c356f368c9950ae0d0deb

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 3]
```diff
@@ -1398,6 +1465,255 @@ def build_conversion_tasks(
...
 
+    def _detect_fp8_params(
+        self,
+        megatron_model: List[MegatronModel],
+        model_config: TransformerConfig,
+        sorted_global_param_names_all_pp_ranks: List[str],
+        pp_group: Any,
+        fp8_scale_inv_attr: str,
+    ) -> Dict[str, bool]:
+        """Detect which global parameters are blockwise FP8 and gather flags across pipeline parallel ranks.
```
Comment (cuichenx) [inline src/megatron/bridge/models/conversion/model_bridge.py:1499]: does this not work with tensorwise fp8? if so we should guard it

**Comment by cuichenx** [comment]
I'm curious how we should test this. Would the `hf_megatron_roundtrip.py` script for a model like [Qwen/Qwen3-4B-Instruct-2507-FP8](https://huggingface.co/Qwen/Qwen3-4B-Instruct-2507-FP8) work after this PR? Or does it require a checkpoint trained in FP8? If you could comment on what type of FP8 configs you verified (and which configs you think would not work) that would be great

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 3 above)
Comment (eternally-z) [reply src/megatron/bridge/models/conversion/model_bridge.py:1499]: > does this not work with tensorwise fp8? if so we should guard it

Yes, that is right. Currently, this implementation only supports blockwise FP8. I will add a guard check to explicitly handle this and update the PR soon. Thanks for pointing it out!

**Comment by eternally-z** [comment]
> I'm curious how we should test this. Would the `hf_megatron_roundtrip.py` script for a model like [Qwen/Qwen3-4B-Instruct-2507-FP8](https://huggingface.co/Qwen/Qwen3-4B-Instruct-2507-FP8) work after this PR? Or does it require a checkpoint trained in FP8? If you could comment on what type of FP8 configs you verified (and which configs you think would not work) that would be great

Regarding the testing, it does not require a trained checkpoint in FP8. The conversion logic works as long as the Megatron configuration has the corresponding FP8 flags enabled during the process.
Verified/Supported Configuration:
- fp8: 'e4m3' or 'hybrid'
- fp8_recipe: 'blockwise'
- fp8_param(fp8_param_gather): True 

Unsupported Configurations:
Other configurations, such as the delayed scaling (tensorwise) recipe, are currently not supported. I will update the PR with the guarding code soon so that unsupported configs will be caught explicitly.

**Comment by eternally-z** [comment]
Hi, @cuichenx
I have updated the PR with the guarding logic as discussed.
Specifically, I added a helper function `_validate_fp8_export_config` which is called within the export_hf_weight path. It ensures that the current Megatron configuration is compatible with the FP8 export requirements:
- fp8 must be enabled.
- fp8_recipe must be set to 'blockwise'.
- fp8_param must be True 

Any other configurations (such as delayed scaling / tensorwise) will now raise a descriptive ValueError to prevent incorrect exports.

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1994/commits/53c4d1dcfcef1e8ab833e594cf2d7717c08e6eb6

**Comment by yaoyu-33** [comment]
@eternally-z your tests failed.

**Comment by eternally-z** [comment]
> @eternally-z your tests failed.

@yaoyu-33  Sorry about that. I have fixed the tests in the latest commit. Could you please help re-trigger the CI?

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1994/commits/f05fffc3b34c2f4474e02ebe962a462d63741494

**Comment by eternally-z** [comment]
@cuichenx hi,I noticed the Codecov patch coverage was below the 80% target, so I have refactored the test file to improve it.  Could you please trigger the CI again?

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1994/commits/af859d8ce04aa362f75d18b75764d08f3495fb90

**Comment by eternally-z** [comment]
@cuichenx hi, the CI failed due to a timeout. The environment setup (building transformer-engine) consumed most of the time limit, leaving insufficient time for the tests to complete. It seems this failure is unrelated to the logic or changes in this PR. Could you please help re-run the CI?

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1994/commits/53d747578d8e2f7130f43fe67cc449f55a89a1e5
